File: .agent/skills/self-evolution/scripts/learning_status.py
#!/usr/bin/env python3
"""
Learning Status Display for SelfEvolution v4.0
Shows active mistakes and improvements before coding
"""

import yaml
import json
import sys
from pathlib import Path
from typing import Dict, List, Optional

# Add scripts directory to path
SCRIPT_DIR = Path(__file__).parent
sys.path.insert(0, str(SCRIPT_DIR))

from project_utils import (
    get_mistakes_file,
    get_improvements_file,
    get_meta_file,
    detect_version
)
from event_tracker import get_statistics

def load_active_mistakes() -> List[Dict]:
    """Load active mistakes (not deprecated)"""
    mistakes_file = get_mistakes_file()
    
    if not mistakes_file.exists():
        return []
    
    with open(mistakes_file, 'r', encoding='utf-8') as f:
        data = yaml.safe_load(f) or {}
    
    mistakes = data.get('mistakes', [])
    return [m for m in mistakes if m.get('status', 'active') == 'active']

def load_active_improvements() -> List[Dict]:
    """Load active improvements (not deprecated)"""
    improvements_file = get_improvements_file()
    
    if not improvements_file.exists():
        return []
    
    with open(improvements_file, 'r', encoding='utf-8') as f:
        data = yaml.safe_load(f) or {}
    
    improvements = data.get('improvements', [])
    return [i for i in improvements if i.get('status', 'active') == 'active']

def format_mistake_short(mistake: Dict) -> str:
    """Format mistake for summary display"""
    lesson = mistake.get('lesson', '').split('.')[0]  # First sentence
    if len(lesson) > 60:
        lesson = lesson[:57] + '...'
    
    version = f"v{mistake.get('version', 1)}"
    return f"❌ {mistake['id']} {version}: {lesson}"

def format_improvement_short(improvement: Dict) -> str:
    """Format improvement for summary display"""
    imp_text = improvement.get('improvement', '').split('.')[0]
    if len(imp_text) > 60:
        imp_text = imp_text[:57] + '...'
    
    version = f"v{improvement.get('version', 1)}"
    return f"✅ {improvement['id']} {version}: {imp_text}"

def display_status(detailed: bool = False):
    """
    Display current learning status
    
    Args:
        detailed: If True, show full details. Otherwise show summary.
    """
    version = detect_version()
    
    if version != '4.0':
        print(f"⚠️  SelfEvolution v{version} detected")
        print("   Run migration to v4.0 first")
        return
    
    stats = get_statistics()
    mistakes = load_active_mistakes()
    improvements = load_active_improvements()
    
    # Header
    print("┌" + "─" * 60 + "┐")
    print("│ 🧠 Auto-Learning – Active for this Project" + " " * 13 + "│")
    print("├" + "─" * 60 + "┤")
    
    # Self-improve status
    if stats['improve_count'] > 0:
        print(f"│ ⚡ Skill Self-Improved ({stats['improve_count']} times)" + " " * 30 + "│")
        print(f"│   Last improved: {stats['last_improved'][:10] if stats['last_improved'] else 'Never'}" + " " * 33 + "│")
        print("│" + " " * 60 + "│")
    
    # Event counter
    threshold_status = f"{stats['event_counter']['since_last_improve']}/{stats['threshold']}"
    next_trigger = stats.get('next_trigger', 5)
    print(f"│ 📊 Events: {stats['event_counter']['total']} total, {threshold_status} to next improve (at #{next_trigger})" + " " * (35 - len(threshold_status) - len(str(stats['event_counter']['total'])) - len(str(next_trigger))) + "│")
    print("│" + " " * 60 + "│")
    
    # Mistakes
    print(f"│ @Mistakes ({len(mistakes)} active)" + " " * (60 - 17 - len(str(len(mistakes)))) + "│")
    
    if mistakes:
        for mistake in mistakes[:5]:  # Show max 5
            lesson = mistake.get('lesson', '')[:50]
            version_info = f"v{mistake.get('version', 1)}"
            print(f"│   ❌ {mistake['id']} {version_info}: {lesson}" + " " * (60 - 8 - len(mistake['id']) - len(version_info) - len(lesson)) + "│")
        
        if len(mistakes) > 5:
            print(f"│   ... and {len(mistakes) - 5} more" + " " * (60 - 17 - len(str(len(mistakes) - 5))) + "│")
    else:
        print("│   (none)" + " " * 51 + "│")
    
    print("│" + " " * 60 + "│")
    
    # Improvements
    print(f"│ @Improvements ({len(improvements)} active)" + " " * (60 - 21 - len(str(len(improvements)))) + "│")
    
    if improvements:
        for improvement in improvements[:5]:
            imp_text = improvement.get('improvement', '')[:47]
            version_info = f"v{improvement.get('version', 1)}"
            print(f"│   ✅ {improvement['id']} {version_info}: {imp_text}" + " " * (60 - 8 - len(improvement['id']) - len(version_info) - len(imp_text)) + "│")
        
        if len(improvements) > 5:
            print(f"│   ... and {len(improvements) - 5} more" + " " * (60 - 17 - len(str(len(improvements) - 5))) + "│")
    else:
        print("│   (none)" + " " * 51 + "│")
    
    print("│" + " " * 60 + "│")
    print("│ ℹ️  Applied to coding decisions via query_lessons.py" + " " * 8 + "│")
    print("└" + "─" * 60 + "┘")
    
    # Detailed view
    if detailed and (mistakes or improvements):
        print("\n" + "=" * 60)
        print("DETAILED VIEW")
        print("=" * 60 + "\n")
        
        if mistakes:
            print("@Mistakes:")
            for mistake in mistakes:
                print(f"\n{mistake['id']} (v{mistake.get('version', 1)}, {mistake.get('severity', 'medium')})")
                print(f"  Problem: {mistake.get('problem', 'N/A')}")
                print(f"  Lesson: {mistake.get('lesson', 'N/A')}")
                print(f"  Stats: {mistake.get('hitCount', 0)} hits, {mistake.get('appliedCount', 0)} applied")
        
        if improvements:
            print("\n@Improvements:")
            for improvement in improvements:
                print(f"\n{improvement['id']} (v{improvement.get('version', 1)})")
                print(f"  Improvement: {improvement.get('improvement', 'N/A')}")
                print(f"  Benefit: {improvement.get('benefit', 'N/A')}")
                print(f"  Stats: {improvement.get('appliedCount', 0)} applied")

def display_versions():
    """Display version history"""
    from project_utils import get_versions_dir
    
    versions_dir = get_versions_dir()
    
    if not versions_dir.exists():
        print("No version history found")
        return
    
    # List all version files
    mistake_versions = sorted(versions_dir.glob('mistakes-v*.yaml'))
    improvement_versions = sorted(versions_dir.glob('improvements-v*.yaml'))
    
    print("📚 Version History\n")
    
    if mistake_versions:
        print("Mistakes:")
        for v_file in mistake_versions:
            version = v_file.stem.split('-v')[1]
            size = v_file.stat().st_size
            modified = v_file.stat().st_mtime
            from datetime import datetime
            modified_str = datetime.fromtimestamp(modified).strftime('%Y-%m-%d %H:%M')
            print(f"  v{version}: {modified_str} ({size} bytes)")
    
    if improvement_versions:
        print("\nImprovements:")
        for v_file in improvement_versions:
            version = v_file.stem.split('-v')[1]
            size = v_file.stat().st_size
            modified = v_file.stat().st_mtime
            from datetime import datetime
            modified_str = datetime.fromtimestamp(modified).strftime('%Y-%m-%d %H:%M')
            print(f"  v{version}: {modified_str} ({size} bytes)")

def display_history():
    """Display self-improve history"""
    meta_file = get_meta_file()
    
    if not meta_file.exists():
        print("No history found")
        return
    
    with open(meta_file, 'r', encoding='utf-8') as f:
        meta = json.load(f)
    
    history = meta.get('self_improve', {}).get('history', [])
    
    if not history:
        print("No self-improve cycles yet")
        return
    
    print("🔄 Self-Improve History\n")
    
    for entry in history:
        cycle = entry.get('cycle_number', '?')
        triggered_at = entry.get('triggered_at', '')[:19]
        changes = entry.get('changes', {})
        
        print(f"Cycle #{cycle} - {triggered_at}")
        print(f"  Mistakes: {changes.get('mistakes_refined', 0)} refined, {changes.get('mistakes_added', 0)} added, {changes.get('mistakes_deprecated', 0)} deprecated")
        print(f"  Improvements: {changes.get('improvements_refined', 0)} refined, {changes.get('improvements_added', 0)} added, {changes.get('improvements_deprecated', 0)} deprecated")
        print()

def main():
    """CLI entry point"""
    import argparse
    
    parser = argparse.ArgumentParser(description='Learning Status Display')
    parser.add_argument('command', nargs='?', default='status', 
                       choices=['status', 'versions', 'history'],
                       help='Command to run')
    parser.add_argument('--detailed', '-d', action='store_true',
                       help='Show detailed view')
    
    args = parser.parse_args()
    
    if args.command == 'status':
        display_status(detailed=args.detailed)
    elif args.command == 'versions':
        display_versions()
    elif args.command == 'history':
        display_history()

if __name__ == "__main__":
    main()

File: .agents/skills/self-evolution/scripts/learning_status.js
#!/usr/bin/env node
/**
 * Learning Status Display for SelfEvolution v4.0
 * Shows active mistakes and improvements before coding
 */

import fs from 'fs';
import yaml from 'js-yaml';
import { fileURLToPath } from 'url';
import {
    getMistakesFile,
    getImprovementsFile,
    getMetaFile,
    getVersionsDir,
    detectVersion
} from './project_utils.js';
import { getStatistics } from './event_tracker.js';

const __filename = fileURLToPath(import.meta.url);

/**
 * Load active mistakes (not deprecated)
 */
export function loadActiveMistakes() {
    try {
        const mistakesFile = getMistakesFile();
        if (!fs.existsSync(mistakesFile)) return [];

        const data = yaml.load(fs.readFileSync(mistakesFile, 'utf-8')) || {};
        const mistakes = data.mistakes || [];
        return mistakes.filter(m => (m.status || 'active') === 'active');
    } catch (e) {
        return [];
    }
}

/**
 * Load active improvements (not deprecated)
 */
export function loadActiveImprovements() {
    try {
        const improvementsFile = getImprovementsFile();
        if (!fs.existsSync(improvementsFile)) return [];

        const data = yaml.load(fs.readFileSync(improvementsFile, 'utf-8')) || {};
        const improvements = data.improvements || [];
        return improvements.filter(i => (i.status || 'active') === 'active');
    } catch (e) {
        return [];
    }
}

/**
 * Display current learning status
 */
export function displayStatus(detailed = false) {
    const version = detectVersion();

    if (version !== '4.0') {
        console.log(`⚠️  SelfEvolution v${version} detected`);
        console.log('   Run migration to v4.0 first');
        return;
    }

    const stats = getStatistics();
    const mistakes = loadActiveMistakes();
    const improvements = loadActiveImprovements();

    // Header
    console.log('┌' + '─'.repeat(60) + '┐');
    console.log('│ 🧠 Auto-Learning – Active for this Project' + ' '.repeat(13) + '│');
    console.log('├' + '─'.repeat(60) + '┤');

    // Self-improve status
    if (stats.improve_count > 0) {
        const lastImproved = stats.last_improved ? stats.last_improved.slice(0, 10) : 'Never';
        console.log(`│ ⚡ Skill Self-Improved (${stats.improve_count} times)`.padEnd(61) + '│');
        console.log(`│   Last improved: ${lastImproved}`.padEnd(61) + '│');
        console.log('│' + ' '.repeat(60) + '│');
    }

    // Event counter
    const thresholdStatus = `${stats.event_counter.since_last_improve}/${stats.threshold}`;
    console.log(`│ 📊 Events: ${stats.event_counter.total} total, ${thresholdStatus} to next improve`.padEnd(61) + '│');
    console.log('│' + ' '.repeat(60) + '│');

    // Mistakes
    console.log(`│ @Mistakes (${mistakes.length} active)`.padEnd(61) + '│');

    if (mistakes.length > 0) {
        for (const mistake of mistakes.slice(0, 5)) {
            const lesson = (mistake.lesson || '').slice(0, 45);
            const vInfo = `v${mistake.version || 1}`;
            console.log(`│   ❌ ${mistake.id} ${vInfo}: ${lesson}`.padEnd(61) + '│');
        }
        if (mistakes.length > 5) {
            console.log(`│   ... and ${mistakes.length - 5} more`.padEnd(61) + '│');
        }
    } else {
        console.log('│   (none)'.padEnd(61) + '│');
    }

    console.log('│' + ' '.repeat(60) + '│');

    // Improvements
    console.log(`│ @Improvements (${improvements.length} active)`.padEnd(61) + '│');

    if (improvements.length > 0) {
        for (const improvement of improvements.slice(0, 5)) {
            const impText = (improvement.improvement || '').slice(0, 42);
            const vInfo = `v${improvement.version || 1}`;
            console.log(`│   ✅ ${improvement.id} ${vInfo}: ${impText}`.padEnd(61) + '│');
        }
        if (improvements.length > 5) {
            console.log(`│   ... and ${improvements.length - 5} more`.padEnd(61) + '│');
        }
    } else {
        console.log('│   (none)'.padEnd(61) + '│');
    }

    console.log('│' + ' '.repeat(60) + '│');
    console.log('│ ℹ️  Applied to coding decisions via query_lessons.js' + ' '.repeat(6) + '│');
    console.log('└' + '─'.repeat(60) + '┘');

    // Detailed view
    if (detailed && (mistakes.length > 0 || improvements.length > 0)) {
        console.log('\n' + '='.repeat(60));
        console.log('DETAILED VIEW');
        console.log('='.repeat(60) + '\n');

        if (mistakes.length > 0) {
            console.log('@Mistakes:');
            for (const mistake of mistakes) {
                console.log(`\n${mistake.id} (v${mistake.version || 1}, ${mistake.severity || 'medium'})`);
                console.log(`  Problem: ${mistake.problem || 'N/A'}`);
                console.log(`  Lesson: ${mistake.lesson || 'N/A'}`);
                console.log(`  Stats: ${mistake.hitCount || 0} hits, ${mistake.appliedCount || 0} applied`);
            }
        }

        if (improvements.length > 0) {
            console.log('\n@Improvements:');
            for (const improvement of improvements) {
                console.log(`\n${improvement.id} (v${improvement.version || 1})`);
                console.log(`  Improvement: ${improvement.improvement || 'N/A'}`);
                console.log(`  Benefit: ${improvement.benefit || 'N/A'}`);
                console.log(`  Stats: ${improvement.appliedCount || 0} applied`);
            }
        }
    }
}

/**
 * Display version history
 */
export function displayVersions() {
    const versionsDir = getVersionsDir();

    if (!fs.existsSync(versionsDir)) {
        console.log('No version history found');
        return;
    }

    const files = fs.readdirSync(versionsDir);
    const mistakeVersions = files.filter(f => f.startsWith('mistakes-v')).sort();
    const improvementVersions = files.filter(f => f.startsWith('improvements-v')).sort();

    console.log('📚 Version History\n');

    if (mistakeVersions.length > 0) {
        console.log('Mistakes:');
        for (const vFile of mistakeVersions) {
            const version = vFile.split('-v')[1].replace('.yaml', '');
            const stats = fs.statSync(path.join(versionsDir, vFile));
            const modified = new Date(stats.mtime).toISOString().slice(0, 16).replace('T', ' ');
            console.log(`  v${version}: ${modified} (${stats.size} bytes)`);
        }
    }

    if (improvementVersions.length > 0) {
        console.log('\nImprovements:');
        for (const vFile of improvementVersions) {
            const version = vFile.split('-v')[1].replace('.yaml', '');
            const stats = fs.statSync(path.join(versionsDir, vFile));
            const modified = new Date(stats.mtime).toISOString().slice(0, 16).replace('T', ' ');
            console.log(`  v${version}: ${modified} (${stats.size} bytes)`);
        }
    }
}

/**
 * Display self-improve history
 */
export function displayHistory() {
    try {
        const metaFile = getMetaFile();
        if (!fs.existsSync(metaFile)) {
            console.log('No history found');
            return;
        }

        const meta = JSON.parse(fs.readFileSync(metaFile, 'utf-8'));
        const history = (meta.self_improve || {}).history || [];

        if (history.length === 0) {
            console.log('No self-improve cycles yet');
            return;
        }

        console.log('🔄 Self-Improve History\n');

        for (const entry of history) {
            const cycle = entry.cycle_number || '?';
            const triggeredAt = (entry.triggered_at || '').slice(0, 19);
            const changes = entry.changes || {};

            console.log(`Cycle #${cycle} - ${triggeredAt}`);
            console.log(`  Mistakes: ${changes.mistakes_refined || 0} refined, ${changes.mistakes_added || 0} added, ${changes.mistakes_deprecated || 0} deprecated`);
            console.log(`  Improvements: ${changes.improvements_refined || 0} refined, ${changes.improvements_added || 0} added, ${changes.improvements_deprecated || 0} deprecated`);
            console.log();
        }
    } catch (e) {
        console.log('No history found');
    }
}

function main() {
    const args = process.argv.slice(2);
    const command = args[0] || 'status';
    const detailed = args.includes('--detailed') || args.includes('-d');

    switch (command) {
        case 'status':
            displayStatus(detailed);
            break;
        case 'versions':
            displayVersions();
            break;
        case 'history':
            displayHistory();
            break;
        default:
            console.log('Learning Status Display');
            console.log('\nCommands:');
            console.log('  status    Show learning status (default)');
            console.log('  versions  Show version history');
            console.log('  history   Show self-improve history');
            console.log('\nOptions:');
            console.log('  --detailed, -d  Show detailed view');
    }
}

if (process.argv[1] === __filename) {
    main();
}
